pydis_site/apps/forms/authentication.py

```diff
@@ -0,0 +1,167 @@
+"""Custom authentication for the forms backend."""
+
+import typing
+
+import jwt
+from django.conf import settings
+from django.http import HttpRequest
+from rest_framework.authentication import BaseAuthentication
+from rest_framework.exceptions import AuthenticationFailed
+
+from . import discord
+from . import models
+
+
+def encode_jwt(info: dict, *, signing_secret_key: str = settings.SECRET_KEY) -> str:
+    """Encode JWT information with either the configured signing key or a passed one."""
+    return jwt.encode(info, signing_secret_key, algorithm="HS256")
+
+
+class FormsUser:
+    """Stores authentication information for a forms user."""
+
+    # This allows us to safely use the same checks that we could use on a Django user.
+    is_authenticated: bool = True
+
+    def __init__(
+        self,
+        token: str,
+        payload: dict[str, typing.Any],
+        member: models.DiscordMember | None,
+    ) -> None:
+        """Set up a forms user."""
+        self.token = token
+        self.payload = payload
+        self.admin = False
+        self.member = member
+
+    @property
+    def display_name(self) -> str:
+        """Return username and discriminator as display name."""
+        return f"{self.payload['username']}#{self.payload['discriminator']}"
+
+    @property
+    def discord_mention(self) -> str:
+        """Return a mention for this user on Discord."""
+        return f"<@{self.payload['id']}>"
+
+    @property
+    def user_id(self) -> str:
+        """Return this user's ID as a string."""
+        return str(self.payload["id"])
+
+    @property
+    def decoded_token(self) -> dict[str, any]:
+        """Decode the information stored in this user's JWT token."""
+        return jwt.decode(self.token, settings.SECRET_KEY, algorithms=["HS256"])
+
+    def get_roles(self) -> tuple[str, ...]:
+        """Get a tuple of the user's discord roles by name."""
+        if not self.member:
+            return []
+
+        server_roles = discord.get_roles()
+        roles = [role.name for role in server_roles if role.id in self.member.roles]
+
+        if "admin" in roles:
+            # Protect against collision with the forms admin role
+            roles.remove("admin")
+            roles.append("discord admin")
+
+        return tuple(roles)
+
+    def is_admin(self) -> bool:
+        """Return whether this user is an administrator."""
+        self.admin = models.Admin.objects.filter(id=self.payload["id"]).exists()
```

Comment:
Should `self.admin` be "private"? Currently, if someone inspects `self.admin` directly without calling `is_admin()`, its value may not be correct. The requirement to call `is_admin()` first is not obvious.